lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -903,20 +903,28 @@ static bool checkConcatNodeUniformDims(llvm::ArrayRef<NodeValue> inputs,
   return true;
 }
 
-// Given one tensor and the desired sub-tensor size, \returns the dimension,
-// after which the sub-tensor has the desired size, otherwise return -1.
-// Example: Given a tensor <2,3,4,5>, and a desired size 20, this function will
-// return dimension 1 as the sub-tensor after it is <4,5>, which has the
-// size 20.
-static ssize_t
-findMatchingConcatDimForSameSizeSubTensor(llvm::ArrayRef<size_t> firstDims,
-                                          size_t desiredSubTensorSize) {
-  size_t currentSubTensorSize = 1;
+/// Given one tensor and the desired leading/trailing dimensions sizes.
+/// \returns the dimension, at which the trailing/leading dimensions match the
+/// desired sizes, otherwise returns -1.
+/// Example: Given a tensor <1,2,3,4,5>, and a desired trailing dimensions size
+/// of 20, and a desired leading dimensions size of 2, this function will return
+/// dimension 1 as the trailing dimensions after it are <4,5>, which matches the
+/// size 20, and the leading dimensions are <1,2>, which matches the size 2.
+static ssize_t findMatchingConcatDimForSameSizeSubTensor(
+    llvm::ArrayRef<size_t> firstDims, size_t leadingDimsProdOriginalConcatNode,
+    size_t trailingDimsProdOriginalConcatNode) {
+  size_t trailingDimsProdCurNode = 1;
   for (ssize_t i = firstDims.size() - 1; i >= 0; i--) {
-    if (currentSubTensorSize == desiredSubTensorSize) {
-      return i;
+    if (trailingDimsProdCurNode == trailingDimsProdOriginalConcatNode) {
+      size_t leadingDimsProdCurNode = 1;
+      for (size_t j = 0; j < i; j++) {
```

Comment:
It's not necessary for this to be a nested loop; I'd prefer it to be refactored into two non-nested loops to be more readable. The first loop looks for `i` where `trailingDimsProdCurNode == trailingDimsProdOriginalConcatNode`. If it doesn't find such an `i` then it returns `-1`. The second loop is then essentially then the inner loop here, given this `i`.

Reply:
Thx for the advice! However it's possible that there are multiple i where `trailingDimsProdCurNode == trailingDimsProdOriginalConcatNode`. 
say for example like <2,3,1,1,1,1,3>, with `trailingDimsProdOriginalConcatNode = 3`, `leadingDimsProdOriginalConcatNode=2`, if we early break the first loop when i = 6, then the current leading dims size is 6, which is not desired. But if we can iterate to i=1, we then can return the correct dim.

Reply:
Ah ok cool got it. I suggest adding a unit test for this case then, as it seems this is currently not covered by the unit test you added.